share/wake/lib/system/sources.wake

```diff
@@ -58,6 +58,68 @@ export def source (file: String): Result Path Error =
         x, Nil = Pass x
         _      = unreachable "a quoted RegExp can never match more than one distinct String"
 
+# Allows claiming of a file so long as the file is within
+# wake's workspace. Keep in mind that if another job
+# produced this file in the same run of wake, claim will
+# fail. Likewise if you claim a file then any job that
+# produces this file will fail. Claim should be thought
+# of as a "virtual" job and wake demands that all paths
+# be produced by exactly one job, virtual or otherwise.
+#
+# For source files please use `source`. For files outside
+# of wake's workspace please use `claimFileAsPath` or
+# `claimFileAsPathIn`. `claim` is good for claiming
+# artifacts of previous builds that should not be rebuilt
+# during this run.
+#
+# Parameters:
+#  - `file`: The string giving the path to the file to be claimed
+#
+# Example:
+# ```
+#   claim "build/my_flow/my_step/previous-output.txt" # Claims the file
+# ```
+export def claim (file: String): Result Path Error =
+    # Define the actual virtual job that produces the Path
+    def raw_claim file =
+        def get_modtime file = prim "get_modtime"
+        def time = get_modtime file
+        if time == -1 then
+            failWithError "{file}: file does not exist. Maybe you haven't previously built this file?"
+        else
+            # Its important that the command here is distinct from what `source` would use
+            # so that claim and source do not overlap.
+            makeExecPlan ("<claim>", str time, file, Nil) Nil
+            | setPlanShare       False
+            | setPlanEcho        logDebug
+            | setPlanEnvironment Nil
+            | setPlanFnOutputs   (file, _)
+            | runJobWith virtualRunner
+            | getJobOutput
+
+    # Compute the relative path in simplest form
+    def isAbs = matches `/.*` file
+    def relPath = if isAbs then
+        relative workspace file
+    else
+        simplify file
+
+    # Fail if we're outside of wake's workspace (relPath must be simplified)
+    def isOutsideRoot = matches `\.\..*/.*` relPath
```

Comment:
I missed this regex in this version of the PR.  The first `.*` needs to be removed; as-is, it rejects `..x/y` which is a perfectly legal -- if not recommended -- file within the workspace, while allowing `..` without a trailing slash (until that gets caught by the "not a file" check) later.  The former is something we should allow regardless; for the latter I'd argue that `..` and `../` are conceptually identical and shouldn't be treated differently -- especially when `isOutsideRoot`, in isolation, will sometimes say that the parent directory is actually located within the root.

Reply:
Ah indeed ..x/y is rejected! This is why I don't like using regexes for this stuff! 

hmm whats the right regex now?

`\.\./.*` is what I think your suggesting now? I'll try that out (seems right at a glance to me). We should probably add better file path handling to wake as primitives.

Reply:
Yeah, that's what would fix `claim`.  If you want to unify `..` and `../`, that would be `\.\.(/.*)?`, but I know you aren't as fond of that one.